File: apps/java/src/main/java/chapter03/Main.java
package chapter03;

import chapter01.KinokoTakenoko;
import chapter02.Features;
import chapter02.Preprocessing;
import chapter02.TrainTestSplit;
import dataset.DataDir;
import java.io.IOException;
import java.util.List;
import java.util.Locale;

/** 深さごとの正解率と、深さ 2 の決定木を表示する。 */
public final class Main {
  private static final double TEST_SIZE = 0.3;
  private static final long SEED = 0;
  private static final List<Integer> MAX_DEPTHS = List.of(1, 2, 3, 4, 5);
  private static final int TREE_DEPTH_TO_SHOW = 2;

  private Main() {}

  public static void main(String[] args) throws IOException {
    TrainTestSplit<Features, String> split =
        Preprocessing.prepareIris(DataDir.dataDir().resolve("iris.csv"), TEST_SIZE, SEED);
    System.out.println("深さ\t訓練データ\tテストデータ");
    for (int maxDepth : MAX_DEPTHS) {
      printAccuracy(String.valueOf(maxDepth), DecisionTree.withMaxDepth(maxDepth), split);
    }
    printAccuracy("制限なし", DecisionTree.unlimited(), split);

    DecisionTree shallow =
        DecisionTree.withMaxDepth(TREE_DEPTH_TO_SHOW).fit(split.xTrain(), split.tTrain());
    System.out.println();
    System.out.println("深さ " + TREE_DEPTH_TO_SHOW + " の決定木:");
    System.out.println(DecisionTrees.format(shallow.tree().orElseThrow()));
  }

  private static void printAccuracy(
      String label, DecisionTree model, TrainTestSplit<Features, String> split) {
    model.fit(split.xTrain(), split.tTrain());
    double train = KinokoTakenoko.accuracy(model.predict(split.xTrain()), split.tTrain());
    double test = KinokoTakenoko.accuracy(model.predict(split.xTest()), split.tTest());
    System.out.println(label + "\t" + format(train) + "\t" + format(test));
  }

  private static String format(double value) {
    return String.format(Locale.ROOT, "%.4f", value);
  }
}

File: apps/kotlin/src/main/kotlin/chapter03/Main.kt
package chapter03

import chapter01.accuracy
import chapter02.prepareIris
import dataset.dataDir
import java.io.File
import java.util.Locale

private const val TEST_SIZE = 0.3
private const val SEED = 0
private val MAX_DEPTHS = listOf(1, 2, 3, 4, 5, null)
private const val TREE_DEPTH_TO_SHOW = 2

private fun format(value: Double): String = "%.4f".format(Locale.ROOT, value)

fun main() {
    val split = prepareIris(File(dataDir(), "iris.csv"), testSize = TEST_SIZE, seed = SEED)
    println("深さ\t訓練データ\tテストデータ")
    for (maxDepth in MAX_DEPTHS) {
        val model = DecisionTree(maxDepth).fit(split.xTrain, split.tTrain)
        val train = accuracy(model.predict(split.xTrain), split.tTrain)
        val test = accuracy(model.predict(split.xTest), split.tTest)
        println("${maxDepth ?: "制限なし"}\t${format(train)}\t${format(test)}")
    }

    val shallow = DecisionTree(TREE_DEPTH_TO_SHOW).fit(split.xTrain, split.tTrain)
    shallow.tree?.let {
        println("\n深さ $TREE_DEPTH_TO_SHOW の決定木:")
        println(formatTree(it))
    }
}
